drivers/i2s/i2s_cavs.c

```diff
@@ -105,182 +87,141 @@ struct i2s_cavs_dev_data {
 #define DEV_DATA(dev) \
 	((struct i2s_cavs_dev_data *const)(dev)->driver_data)
 
+static struct device DEVICE_NAME_GET(i2s1_cavs);
+static struct device DEVICE_NAME_GET(i2s2_cavs);
+static struct device DEVICE_NAME_GET(i2s3_cavs);
+
 static struct device *get_dev_from_dma_channel(u32_t dma_channel);
-static void dma_tx_callback(struct device *, u32_t, int);
-static void tx_stream_disable(struct stream *,
-		volatile struct i2s_cavs_ssp *const, struct device *);
 
-static inline u16_t modulo_inc(u16_t val, u16_t max)
-{
-	val++;
-	return (val < max) ? val : 0;
-}
+static void i2s_dma_tx_callback(struct device *, u32_t, int);
+static void i2s_tx_stream_disable(struct i2s_cavs_dev_data *,
+		volatile struct i2s_cavs_ssp *const, struct device *);
+static void i2s_rx_stream_disable(struct i2s_cavs_dev_data *,
+		volatile struct i2s_cavs_ssp *const, struct device *);
 
-/*
- * Get data from the queue
- */
-static int queue_get(struct ring_buf *rb, u8_t mode, void **mem_block,
-			size_t *size)
+/* This function is executed in the interrupt context */
+static void i2s_dma_tx_callback(struct device *dev_dma, u32_t channel,
+		int status)
 {
-	unsigned int key;
-
-	key = irq_lock();
+	struct device *dev = get_dev_from_dma_channel(channel);
+	const struct i2s_cavs_config *const dev_cfg = DEV_CFG(dev);
+	struct i2s_cavs_dev_data *const dev_data = DEV_DATA(dev);
+	volatile struct i2s_cavs_ssp *const ssp = dev_cfg->regs;
+	struct stream *strm = &dev_data->tx;
+	void *buffer;
+	int ret;
 
-	/* In case of ping-pong mode, the buffers are not freed after
-	 * reading. They are fixed in size. Another thread will populate
-	 * the pong buffer while the ping buffer is being read out and
-	 * vice versa. Hence, we just need to keep reading from buffer0
-	 * (ping buffer) followed by buffer1 (pong buffer) and the same
-	 * cycle continues.
-	 *
-	 * In case of non-ping-pong modes, each buffer is freed after it
-	 * is read. The tail pointer will keep progressing depending upon
-	 * the reads. The head pointer will move whenever there's a write.
-	 * If tail equals head, it would mean we have read everything there
-	 * is and the buffer is empty.
-	 */
-	if (rb->tail == rb->head) {
-		if ((mode & I2S_OPT_PINGPONG) == I2S_OPT_PINGPONG) {
-			/* Point back to the first element */
-			rb->tail = 0;
-		} else {
-			/* Ring buffer is empty */
-			irq_unlock(key);
-			return -ENOMEM;
-		}
+	ret = k_msgq_get(&strm->out_queue, &buffer, K_NO_WAIT);
```

Comment:
> Is this statement really needed? Can this ever happen? Also, apart from this statement, I am not quite sure what's the use of out_queue for Tx case (playback). I feel it can be excluded for Tx case.

I've added some comments above the definition of struct stream. Please check, I hope that it clarifies the usage of in and out queues.

Also see #10948 which when address should help in reducing the number of queues.